### PR TITLE
Fix to #2027 - improve the way we build IsNull expressions for null semantics

### PR DESCRIPTION
**Pull request by maumar**
Problem was that in null semantics, when we test whether expression is null or not, we assumed incorrectly that the expression is null if either of it's parameters is null.
This works for most cases (e.g. functions), but not for all.
Exapmles are COALESCE and CASE-WHEN-THEN-ELSE.

Fix is to refactor the visitor that is used for that purpose, before it was just extracting nullable expressions, now it builds the entire IsNullExpression tree.

Also fixed some minor issues with case expression - we were assuming it would always produce expression of type bool, but it's no longer the case.

### CODE REVIEW

File: src/EntityFramework.Relational/Query/ExpressionTreeVisitors/IsNullExpressionBuildingVisitor.cs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            var inExpression = expression as InExpression;
+            if (inExpression != null)
+            {
+                return expression;
+            }
+
+            var caseExpression = expression as CaseExpression;
+            if (caseExpression != null)
+            {
+                if (caseExpression.When is ConditionalExpression)
```
Comment (maumar) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/IsNullExpressionBuildingVisitor.cs]: should I just inline the VisitConditionalExpression here instead?

**Comment by maumar** [comment]
@anpete

File: src/EntityFramework.Relational/Query/ExpressionTreeVisitors/IsNullExpressionBuildingVisitor.cs (hunk 1 above)
Comment (anpete) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/IsNullExpressionBuildingVisitor.cs]: This seems fine.

File: test/EntityFramework.SqlServer.FunctionalTests/NullSemanticsQuerySqlServerTest.cs [hunk 2]
```diff
@@ -681,6 +727,94 @@ public override void Where_select_not_equal_nullable_with_null_value_parameter()
...
+        }
+
+        public override void Where_not_equal_with_conditional()
+        {
+            base.Where_not_equal_with_conditional();
+
+            Assert.Equal(
+                @"SELECT [e].[Id]
+FROM [NullSemanticsEntity1] AS [e]
+WHERE ((([e].[NullableStringC] <> CASE WHEN ((([e].[NullableStringA] = [e].[NullableStringB] AND ([e].[NullableStringA] IS NOT NULL AND [e].[NullableStringB] IS NOT NULL)) OR ([e].[NullableStringA] IS NULL AND [e].[NullableStringB] IS NULL))) THEN [e].[NullableStringA] ELSE [e].[NullableStringB] END) OR ([e].[NullableStringC] IS NULL OR CASE WHEN ((([e].[NullableStringA] = [e].[NullableStringB] AND ([e].[NullableStringA] IS NOT NULL AND [e].[NullableStringB] IS NOT NULL)) OR ([e].[NullableStringA] IS NULL AND [e].[NullableStringB] IS NULL))) THEN [e].[NullableStringA] ELSE [e].[NullableStringB] END IS NULL)) AND ([e].[NullableStringC] IS NOT NULL OR CASE WHEN ((([e].[NullableStringA] = [e].[NullableStringB] AND ([e].[NullableStringA] IS NOT NULL AND [e].[NullableStringB] IS NOT NULL)) OR ([e].[NullableStringA] IS NULL AND [e].[NullableStringB] IS NULL))) THEN [e].[NullableStringA] ELSE [e].[NullableStringB] END IS NOT NULL))",
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/NullSemanticsQuerySqlServerTest.cs]: It would be great if we could format this better.

**Comment by anpete** [comment]
:shipit:

**Pull request merged**